### PR TITLE
BLTHS PART3 PROBE MODE (Ability to specify that a probe does not need a prepare and finalize action in between individual probe actions)

### PR DESCRIPTION
**Pull request by FanDjango**
This is **part3** of the **BLTOUCH-HIGH-SPEED** PR split (see #2309).

**OVERVIEW**

Some types of probe, like the BLTouch for example, do not need an intermediate STOW/DEPLOY command sequence when moving from one probing point to the next or when performing multiple probes in the same location.

**COMMENTS**

The speed gain to be achieved here is significant if a few other things are also tweaked. For a BLTouch in HIGH-SPEED-MODE, it is advantageous to lift up as fast as possible after a homing or probing trigger, even if the downward movement on homing or probing is slow. Therefore it is a good idea to configure `lift_speed` as high as possible, the same goes for `homing_retract_speed`. Note that both of these are upward limited by `max_z_velocity`, so check that value as well.

**IMPORTANT NOTE:** This "high-speed-mode" uses the original `bltouch.py` module with minimum changes to enable the message passing of DEPLOY and STOW actions by the send_event mechanism. This does increase the probing speed, especially if you also increase the retract speeds on homing and the lift_speed on probing.  

**CONFIG**

If this "**high_speed_mode**" is desired, add a `stow_on_each_sample` parameter to the probes config section. Currently there is only one probe that supports this parameter: the BLTouch - so we are talking about adding this parameter to the `[bltouch]` section.

```
[bltouch]
stow_on_each_sample: False
```

**ADDITIONAL INFORMATION**

The branch associated with PR #2309 contains this code (BLTHS PART3 PROBE MODE) and can be found at [this place](https://github.com/FanDjango/klipper/tree/BLTOUCH-HIGH-SPEED) for all that would like to help test this stuff

Signed-off-by: Mike Stiemke **fandjango@gmx.de**

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks.  The functionality looks useful and I agree it is a worthwhile addition.

I do wonder if we could reduce the complexity in homing.py and probe.py for this feature.  Both of these files are already complex, so if we can keep this feature mostly contained in bltouch.py I think that would be advantageous.

Along those lines, instead of adding "probe:stow_needed"/"probe_deploy_need" events, what if we added ProbeEndstopWrapper:start_multi_probe() and ProbeEndstopWrapper:finish_multi_probe() calls.  Dummy versions could be added to ProbeEndstopWrapper, while BLTouchEndstopWrapper could use these callbacks to track that multiple probes are possible.  Then, in the bltouch version of home_finalize() it could avoid stowing the probe if its internal state indicates there is a multi-probing sequence occurring.  Similarly, its home_prepare() code could skip deploying the probe if the local state indicates it has already been deployed.  Finally, the finish_multi_probes() code could stow the probe if the internal state indicates the probe has been deployed.

Separately, as a minor suggestions, what about using a config option like `stow_on_each_sample: True` instead of `high_speed_mode: False`.

Thoughts?
-Kevin

**Comment by FanDjango** [comment]
I see what you are saying. I will need to look at this. I will work on PART2 HEATERS first.

> I do wonder if we could reduce the complexity in homing.py and probe.py for this feature. Both of these files are already complex, so if we can keep this feature mostly contained in bltouch.py I think that would be advantageous.

Your code, your call. 

Calling it `stow_on_each_sample`? Ok.

**Comment by FanDjango** [comment]
> Then, in the bltouch version of home_finalize() it could avoid stowing the probe if its internal state indicates there is a multi-probing sequence occurring. Similarly, its home_prepare() code could skip deploying the probe if the local state indicates it has already been deployed. Finally, the finish_multi_probes() code could stow the probe if the internal state indicates the probe has been deployed.

I understand that idea. It makes the decision logic (first, mid, last) when to deploy/stow redundant. Very good. Lemme see what I can do...

**Comment by KevinOConnor** [comment]
Thanks!  In general it looks good to me.

When I initially reviewed, I didn't realize you also wanted to do "fast probing" between the first and second homing attempts.  (I thought you were just interested in optimizing the probing sequences in ProbePointsHelper.)  Optimizing homing is fine though.  I do think the code should handle errors that occur during probes and be sure to propagate the finish_multi_probes() call.  I'd expect the call sites (homing, probe_accuracy, probepointshelper) to look something like:
```
    mcu_endstop.start_multi_probes()
    try:
        ...
        mcu_endstop.finish_multi_probes()
    except homing.CommandError:
        logging.exception("Error during multi-probe sequence")
        mcu_endstop.finish_multi_probes()
        raise
```
Otherwise, I fear an error could leave the bltouch in a deployed state, which an inattentive user might miss and go on to command head movement while the probe is deployed.

-Kevin

**Comment by KevinOConnor** [comment]
Actually, on further thought, perhaps a simpler implementation of error handling would be to update gcode.py to invoke `self.printer.send_event("gcode:command_failed")` on an error.  Then the bltouch.py code could register that event and stow the probe if it has been deployed.

-Kevin

**Comment by FanDjango** [comment]
> "fast probing" between the first and second homing attempts

That has always been part of the original intention, same as I did on the "other" firmware. It is not something that "needs" to be done. Because actually the gain is not so big. Therefore, because it intrudes on `mcu_endstop` attributes, I was leery of adding anything to yet another unrelated class. Checking for the presence with `hasattr`, one reads, is less performant that using `try...except` and I thought that this approach might actually be more logical than having a `multi_probe...` definition in `mcu_endstops`. But that is your call.

One reason for also putting this "fast probing" into the homing section: Some day, some person who uses the probe for homing, will ask for this type of homing to do multiple samples and averaging in the same fashion as the real probing can do. And there is no real reason not to support that (some day) when homing, even with an endstop. Once that is done, however, "fast probing" will once again provide a very big speed benefit, so back to square one...

The only real difference between homing and probing (apart from the involved axes) is how the result is used.

> invoke self.printer.send_event("gcode:command_failed") on an error

I understand the thinking behind this. You want me to do that? Wait a moment...

> Then the bltouch.py code could register that event and stow the probe if it has been deployed.

I kind of feel that `bltouch.py` should (for safeties sake) catch **all** and **any** errors and reset the multi-probe state and STOW the probe, regardless of where the error came from, just like it is done on `klippy:connect`. 

The `start_multi_probe()` calls are so close (program-step-wise) to the probing move that in all cases of a STOWED probe on retry or restart of a lower (or higher) level failure, the `start_multi_probe' logic` would redeploy the probe. Especially if we put the `start_multi_probe` logic into **all** places where the probe is being used: into homing (see above), into manual probing or the probe gcode command (which I have done but commented out for now) etc.

> then the bltouch.py code could register that event and stow the probe if it has been deployed.

Just STOW, regardless of wether deployed or not. You will notice that the current implementation has no need to "remember" the pin state, it just tracks wether we are in "multi-probe" sequence of prepare-finalize, ....., prepare-finalize.

So, is there a general error trap somewhere or do you mean I should add a try-except + send_event to all gcode commands invoking probing?

Oh, and by the way, this separated-out PR is now getting very very big itself :-)

**Comment by FanDjango** [comment]
Also, if you agree, can I squash to just have the final version as compared to master here when looking at the diffs, makes it easier to see what is changing and where, if you agree. We don't really need the history of how this PR's code is progressing, do we?

**Comment by KevinOConnor** [comment]
> > invoke self.printer.send_event("gcode:command_failed") on an error

> I understand the thinking behind this. You want me to do that? Wait a moment...

After thinking about this a little more, it occurred to me that homing.py could use some refactoring.  After this change, it will have the following notifications: `mcu_endstop.start_multi_probes()`, `mcu_endstop.home_prepare()`, `send_event("homing:move_begin")`, `mcu_endstop.home_start()`, `mcu_endstop.home_wait()`, `send_event("homing:move_end")`, `mcu_endstop.home_finalize()`, `mcu_endstop.finish_multi_probes()`, and `send_event("homing:homed_rails")`.

I think this could be reworked to: `send_event("homing:start_home_rails")`, `send_event("homing:move_begin")`, `mcu_endstop.home_start()`, `mcu_endstop.home_wait()`, `send_event("homing:move_end")`, and `send_event("homing:homed_rails")`.

That is, I think probe.py could catch new "homing:start_home_rails" / "homing:homed_rails" events and translate them into the probe specific `start_multi_probes()`, `home_prepare()`,  `home_finalize()`, and `finish_multi_probes()` calls.

I can do this refactoring, and I can do it before or after merging (if either is preferable to you).  If you'd prefer to refactor after merging, then we can keep that part of this PR as you have it now.

> So, is there a general error trap somewhere

I was thinking of adding `self.printer.send_event("gcode:command_failed")` to the error handler of gcode.py:_process_commands().

-Kevin

File: klippy/extras/probe.py [hunk 1]
```diff
@@ -346,18 +360,21 @@ def start_probe(self, params):
...
         self.lift_speed = probe.get_lift_speed(params)
         self.probe_offsets = probe.get_offsets()
         if self.horizontal_move_z < self.probe_offsets[2]:
             raise self.gcode.error("horizontal_move_z can't be less than"
                                    " probe's z_offset")
+        probe.start_multi_probes()
         while 1:
-            done = self._move_next()
-            if done:
+            if self._move_next():
```
Comment (KevinOConnor) [inline klippy/extras/probe.py]: This change seems unrelated to the focus of this patch and I'd say it would be better to not mix it in.  (I'm not sure there's a big gain by rewording these lines.

Comment (FanDjango) [reply klippy/extras/probe.py]: Oops, sorry. My brain somehow couldn't resist...

File: klippy/extras/probe.py [hunk 2]
```diff
@@ -311,13 +329,9 @@ def get_lift_speed(self):
...
         toolhead = self.printer.lookup_object('toolhead')
         # Lift toolhead
-        speed = self.lift_speed
-        if not self.results:
-            # Use full speed to first probe position
-            speed = self.speed
         curpos = toolhead.get_position()
         curpos[2] = self.horizontal_move_z
-        toolhead.move(curpos, speed)
+        toolhead.move(curpos, self.lift_speed)
```
Comment (KevinOConnor) [inline klippy/extras/probe.py]: This part of the patch doesn't look correct.  The reason we use the regular moving speed on the first move is for delta printers - the user often starts the probing when the head is very far away from the bed and we want to use the regular move speed (neither lift nor probe) when making the initial move towards the bed.  Also, it seems unrelated to the focus of this PR, so I'd raise in a separate one if a change is needed.

File: klippy/extras/probe.py [hunk 3]
```diff
@@ -346,18 +360,21 @@ def start_probe(self, params):
             self.probe_offsets = (0., 0., 0.)
             self._manual_probe_start()
             return
+        else:
+            self.lift_speed = probe.lift_speed
         # Perform automatic probing
         self.lift_speed = probe.get_lift_speed(params)
```
Comment (KevinOConnor) [inline klippy/extras/probe.py]: Similar for these two lines.  They also look redundant.

**Comment by KevinOConnor** [comment]
> Also, if you agree, can I squash to just have the final version as compared to master here when looking at the diffs, makes it easier to see what is changing and where, if you agree. We don't really need the history of how this PR's code is progressing, do we?

I agree.  I can squash on commit or you can rebase and squash upon making an update.  Either approach is fine.

Thanks,
-Kevin

File: klippy/extras/probe.py (hunk 2 above)
Comment (FanDjango) [reply klippy/extras/probe.py]: Agreed, I didn't realize this was important for deltas. I will undo and fix that. But I will do some testing because I remember that this place was important for some reason while I was searching for the places responsible for my speed observations during intial analysis when probing - I will report on my findings.

File: klippy/extras/probe.py (hunk 3 above)
Comment (FanDjango) [reply klippy/extras/probe.py]: You are an eagle.

**Comment by FanDjango** [comment]
Thanks for reviewing. I will address your non review points here:

> I think this could be reworked to: send_event("homing:start_home_rails"), send_event("homing:move_begin"), mcu_endstop.home_start(), mcu_endstop.home_wait(), send_event("homing:move_end"), and send_event("homing:homed_rails").

Oh yes, much nicer - for homing.py at least. **But**: If we move back to send_event, why not be consistent and use send_event in probing.py as well, as we did originally? It would be analogous to `homing:start_home_rails`: perhaps call it `probing:start_probe_sequence`? Same for `...end...` of course. There are really only three pairs to be inserted into probing.py and then we can forget about the dummy attributes everwhere where they were needed.

> That is, I think probe.py could catch new "homing:start_home_rails" / "homing:homed_rails" events and translate them into the probe specific start_multi_probes(), home_prepare(), home_finalize(), and finish_multi_probes() calls.

Yes.

> I can do this refactoring, and I can do it before or after merging (if either is preferable to you). If you'd prefer to refactor after merging, then we can keep that part of this PR as you have it now.

Depends on your schedule: By when do you anticipate this refactoring to have taken place? On the other hand if you refactor after merging it might be more work for **you** (but I would offer to contribute too of course) **but** I could start implementing your suggestions right away.

>  I can squash on commit or you can rebase and squash upon making an update

Next update, I will make a clean table here.

**Comment by KevinOConnor** [comment]
> If we move back to send_event, why not be consistent and use send_event in probing.py as well, as we did originally? It would be analogous to homing:start_home_rails: perhaps call it probing:start_probe_sequence?

That's possible.  The only two users right now are ProbeEndstopWrapper and BLTouchEndstopWrapper which PrinterProbe already has a direct relationship with.  So, I'd guess it's a little simpler to directly invoke then to use the indirect send_event() mechanism.  It's not a big deal either way.

> By when do you anticipate this refactoring to have taken place? 

It's not a lot of work, so if you'd prefer the refactor first, then I should be able to do it today, and if you prefer after then I'd try to do it within a couple of days after merging this PR.

> As in: You will add that (at some point in time)? I will add that msg receive in bltouch.py in this PR? 

I can take care of that.

Thanks,
-Kevin

**Comment by FanDjango** [comment]
Ok, please refactor first. I would continue on this here on this upcoming monday.

**Comment by FanDjango** [comment]
Ok, I rebased to current, cherry-picked your 4 commits in the PR #2502 and am now working on finishing this. I'll keep you up to date.

**Comment by FanDjango** [comment]
Tested with stock `bltouch.py`: works (as intended by you)
Modified `bltouch.py` with above [commit](https://github.com/KevinOConnor/klipper/pull/2493/commits/183ee7912c56bace99c5ff53cfbf6ca342edfefb): works with parm `stow_on_each_sample: False` and also with `stow_on_each_sample: True` (default).

Tested with **G28 Z**,  **Z_TILT_ADJUST**, **BED_MESH_CALIBRATE** and **PROBE_ACCURACY**.

I would consider (with parm `stow_on_each_sample: False`), the requested feature/option "high-speed-probing" can be seen as finished.

There is one small caveat though:

Look at 

```
    def multi_probe_end(self):
        if self.stow_on_each_sample:
            return
        self.raise_probe()
        self.multi = 'OFF'
```
This is what get's called from `gcode.py` on an error. Trouble is, if the parm is set to disable the feature, the probe will not be raised.

If coded like this:
```
    def multi_probe_end(self):
        self.raise_probe()
        self.multi = 'OFF'
```
it would be ok, but that means a double raise command at the end of a probing sequence (if the feature is disabled).

I would have liked:
```
    def multi_probe_end(self):
        if self.stow_on_each_sample:
            return
        self.raise_probe()
        self.multi = 'OFF'
    def error_probe_end(self):
        self.raise_probe()
        self.multi = 'OFF'
```
Thoughts?

File: klippy/extras/probe.py (hunk 1 above)
Comment (FanDjango) [reply klippy/extras/probe.py]: Obsoleted by refactoring

File: klippy/extras/probe.py (hunk 2 above)
Comment (FanDjango) [reply klippy/extras/probe.py]: Obsoleted by refactoring

File: klippy/extras/probe.py (hunk 3 above)
Comment (FanDjango) [reply klippy/extras/probe.py]: Obsoleted by refactoring

**Comment by KevinOConnor** [comment]
Thanks.  Code looks good to me.  (I'm okay with committing if there are no other concerns.)

> There is one small caveat though:

FWIW, I don't see a code path where the bltouch does not get raised.  (If stow_on_each_sample is True then probe_finalize() should have been called and raised the probe before any error is picked up by the "gcode:command_error" event.)  That said, I've no objections to making the error handling more explicit.  Perhaps a parameter could be added to multi_probe_end - for example: multi_probe_end(did_finish_normally).

-Kevin

**Comment by FanDjango** [comment]
>  a code path where the bltouch does not get raised

I was being paranoid - if an error occurs on the way to the raise.... 

Anyhow thanks for the agreeable code-cooperation.

-Mike

**Comment by KevinOConnor** [comment]
Thanks.  I committed this (commit 086c131b).

Feel free to submit more robust error handling if desired.

-Kevin
